src/main/java/com/conveyal/datatools/manager/jobs/AutoDeployFeedJob.java

```diff
@@ -0,0 +1,133 @@
+package com.conveyal.datatools.manager.jobs;
+
+import com.conveyal.datatools.common.status.MonitorableJob;
+import com.conveyal.datatools.manager.auth.Auth0UserProfile;
+import com.conveyal.datatools.manager.controllers.api.DeploymentController;
+import com.conveyal.datatools.manager.models.Deployment;
+import com.conveyal.datatools.manager.models.FeedVersion;
+import com.conveyal.datatools.manager.models.OtpServer;
+import com.conveyal.datatools.manager.models.Project;
+import com.conveyal.datatools.manager.persistence.Persistence;
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Auto deploy new feed version to OTP server if {@link Project#autoDeploy} is enabled and other conditions are met
+ * (e.g., feed version has no critical errors, active deployment is not in progress, there are no other feed fetches in
+ * progress, etc.). This job must run after {@link ValidateFeedJob} as it has a dependency on the outcome of
+ * {@link FeedVersion#hasCriticalErrors}.
+ */
+public class AutoDeployFeedJob extends MonitorableJob {
+    public static final Logger LOG = LoggerFactory.getLogger(ValidateFeedJob.class);
+
+    /**
+     * Projects to be considered for auto deployment
+     */
+    private List<Project> projects = new ArrayList<>();
+
+    /**
+     * A set of projects which have been locked by a instance of {@link AutoDeployFeedJob} to prevent repeat
+     * deployments.
+     */
+    private final Set<String> lockedProjects = Sets.newConcurrentHashSet();
+
+    /**
+     * Auto deploy specific project.
+     */
+    public AutoDeployFeedJob(Project project, Auth0UserProfile owner) {
+        super(owner, "Auto Deploy Feed", JobType.AUTO_DEPLOY_FEED_VERSION);
+        projects.add(project);
+    }
+
+    /**
+     * Auto deploy all projects.
+     */
+    public AutoDeployFeedJob(Auth0UserProfile owner) {
+        super(owner, "Auto Deploy Feed", JobType.AUTO_DEPLOY_FEED_VERSION);
+        projects = Persistence.projects.getAll();
+    }
+
+    @Override
+    public void jobLogic() {
+        for (Project project : projects) {
+            Deployment deployment = Persistence.deployments.getById(project.pinnedDeploymentId);
+
+            // Define if project and deployment are candidates for auto deploy.
+            if (!project.autoDeploy ||
+                project.pinnedDeploymentId == null ||
+                deployment == null ||
+                deployment.feedVersionIds == null ||
+                lockedProjects.contains(project.id)) {
+                LOG.debug("Project {} skipped for auto deployment as required criteria not met.", project.id);
+                continue;
+            }
+
+            try {
+                LOG.debug("Auto deploy Lock applied to project id: {}", project.id);
+                lockedProjects.add(project.id);
```

Comment:
There is still the possibility of the lockedProjects being modified between the synchronized check for the existence of this project and the actual addition of the project to this set. Therefore, there should be a synchronized block that includes both a check for the existence of the project and then the insertion of the project. Something like this:

```java
synchronized(lockedProjects) {
  if (!lockedProjects.contains(project.id)) {
    lockedProjects.add(project.id);
  }
}
```

Reply:
Updated.